### ISSUE TITLE
Improve image preview speed (MediaStoreThumb vs Exif thumb)

### ISSUE THREAD

**Issue by MFlisar** [bumptech/glide#1285]
I tested loading images via the media store thumbnail vs. getting the exif thumbnail and I think using the exif thumbnail is much faster (resolution is similar, but not equal!).

Media store thubnail size: 512x384
Exif Thumbnail: 215x288

Retrieving exif thumbnail works like following (with sanselan this could be extended to work with streams from any source as well):

```java
public byte[] getExifThumbnail() 
{
	ExifInterface exif;
	try
	{
		exif = new ExifInterface(path);
	}
	catch (IOException e) 
	{
		return null;
	}
	byte[] imageData = exif.getThumbnail();
	if (imageData != null)
		return imageData;
	return null;
}
```

Can I currently use glide to do following:

1) check if exif thumb exists, if yes load it
2) else use media store if possible (this is automatically falling back to the file itself as far as I know already, correct?)

Currently I manually check this condition before creating the glide request, I create different requests depending on if the image has an exif thumbnail.

If I understand glide correctly, when using the media store size or smaller, glide will automatically use the media store thumbnail if it exists and will fall back to loading the image from disc and resize it. I would like to also respect the exif thumbnail as primary thumbnail, then the media store thumbnail and only afterwards fall back to the file...

**Comment by TWiStErRob** [comment]
It only loads the thumb if you `loadFromMediaStore`, otherwise it's always the file. You can write a fetcher/loader that can do what described, however using the exif thumb is an interesting idea and should be considered. The problem I see is that it is too small (not sure if that size is standard). Many of the screens nowadays have very largre resolutions and hence would need much bigger Bitmaps than the exif thumb. This would be only useful for <96dp on xxhdpi which is near icon size.

**Comment by MFlisar** [comment]
So doing following is not enough to get an image from the media store:

    private static final int MEDIA_STORE_THUMB_WIDTH = 512;
    private static final int MEDIA_STORE_THUMB_HEIGHT = 384;
    private static final int MEDIA_STORE_SQUARE = Math.min(MEDIA_STORE_THUMB_HEIGHT, MEDIA_STORE_THUMB_WIDTH);

    // updating requests like following to "force" media store thumbs to be loaded:
    requestBase.override(MEDIA_STORE_SQUARE, MEDIA_STORE_SQUARE);

I have to call use `loadFromMediaStore(uri`)` instead of `load(uri)`? Good to know, I will check that again, maybe this makes things even faster as well...

**Comment by TWiStErRob** [comment]
No, that just loads a similar sized image to media store thumbs.

Answer: https://github.com/bumptech/glide/blob/3.0/library/src/main/java/com/bumptech/glide/load/data/MediaStoreThumbFetcher.java
